### PR TITLE
[SPARK-29022][SQL]Fix SparkSQLCLI can not add jars by AddJarCommand

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
For issue mentioned in [SPARK-29022](https://issues.apache.org/jira/browse/SPARK-29022)
Spark SQL CLI can't use class as serde class in jars add by SQL `ADD JAR`. 

When we create table with `serde` class contains by jar added by SQL 'ADD JAR'.
We can create table with `serde` class construct success since we call `HiveClientImpl.createTable` under `withHiveState` method, it will add `clientLoader.classLoader` to `HiveClientImpl.state.getConf.classLoader`.

Jars added by SQL `ADD JAR` will be add to 
1. `sparkSession.sharedState.jarClassLoader`.
2. 'HiveClientLoader.clientLoader.classLoader'

In Current spark-sql MODE,  `HiveClientImpl.state`  will use CliSessionState created when initialize 
SparkSQLCliDriver,  When we select data from table, it will check `serde` class, when call method `HiveTableScanExec#addColumnMetadataToConf()` to check for table desc serde class. 

```
val deserializer = tableDesc.getDeserializerClass.getConstructor().newInstance()
    deserializer.initialize(hiveConf, tableDesc.getProperties)
```

`getDeserializer` will  use CliSessionState's hiveConf's classLoader in `Spark SQL CLI` mode.
But when we call `ADD JAR` in spark, the jar won't be added to `Classloader of  CliSessionState' conf `, then `ClassNotFound` error happen. 

So we reset `CliSessionState conf's classLoader ` to `sharedState.jarClassLoader` when `sharedState.jarClassLoader` has added jar passed by `HIVEAUXJARS`
Then when we use  `ADD JAR `  to add jar,  jar path will be added to CliSessionState's conf's ClassLoader

### Why are the changes needed?
Fix bug

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
ADD UT

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Hi, @AngersZhuuuu . Is the JIRA issue ID correct? SPARK-29051 ?

**Comment by AngersZhuuuu** [comment]
> Hi, @AngersZhuuuu . Is the JIRA issue ID correct? SPARK-29051 ?

Sorry, I forgot to change it.  This PR should wait for  https://github.com/apache/spark/pull/25542

**Review by srowen — APPROVED** [review]

**Comment by wangyum** [comment]
@dongjoon-hyun @srowen Maybe we can fix this issue by another PR: #25775

**Comment by srowen** [comment]
OK so this should resolve the same issue? are there arguments for against this change vs the other?

**Comment by AngersZhuuuu** [comment]
@srowen @dongjoon-hyun @wangyum 
I have add a new UT for this issue to cover case  `use hive.aux.jars jar's class as serde`. 
#25775  can't solve problem of this issue.

**Comment by AngersZhuuuu** [comment]
@dongjoon-hyun  latest build failed , seems not my change's problem.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 1]
```diff
@@ -144,6 +144,7 @@ private[hive] class HiveClientImpl(
         warehouseDir.foreach { dir =>
           ret.getConf.setVar(ConfVars.METASTOREWAREHOUSE, dir)
         }
+        ret.getConf.setClassLoader(clientLoader.classLoader)
```
Comment (wangyum) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: We need to add comments to explain this change.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala [hunk 2]
```diff
@@ -332,4 +360,18 @@ class CliSuite extends SparkFunSuite with BeforeAndAfterAll with Logging {
...
+
+  test("SPARK-29022 Use add jar class as serde") {
+    val jarFile = HiveTestUtils.getHiveHcatalogCoreJar.getCanonicalPath
+
+    runCliWithin(3.minute)(
+      s"ADD JAR ${jarFile};" -> "",
+      """CREATE TABLE addJar(key string)
+        |ROW FORMAT SERDE 'org.apache.hive.hcatalog.data.JsonSerDe';
+      """.stripMargin -> "",
+      "INSERT OVERWRITE TABLE addJar select 1;" -> "",
```
Comment (wangyum) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: Please add more characters to make test stable. See [SPARK-28988](https://issues.apache.org/jira/browse/SPARK-28988) for more detail.

**Comment by wangyum** [comment]
This PR should test with default behavior, `[test-hadoop3.2]` and `[test-hadoop3.2][test-java11]`.

**Comment by dongjoon-hyun** [comment]
The last python two failures look consistent. Could you take a look if it's relevant?

**Comment by AngersZhuuuu** [comment]
> The last python two failures look consistent. Could you take a look if it's relevant?

Look the test of python, it won't use code I have changed. and you can see the third  build from last  [BUILD](https://github.com/apache/spark/pull/25729#issuecomment-532043555)  and my latest two commit, I just modify UT and add some comment.  Two test result  success no w

**Comment by srowen** [comment]
https://github.com/apache/spark/pull/25775 is merged. I'm not clear if this change is still needed, and if so, if it just needs to be retested?

**Comment by wangyum** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
> #25775 is merged. I'm not clear if this change is still needed, and if so, if it just needs to be retested?

Not same issue, still need this. Checking for reason of build failed.

**Comment by wangyum** [comment]
This PR is needed. but this change can not handle `isolatedLoader` is enabled(`spark.sql.hive.metastore.jars != builtin`). It seems [this commit](https://github.com/apache/spark/pull/25729/commits/0fb149c25e995bd5d2f08ece1ac65ef5f587d722) can handle this case. But we can not add a test case for this case because of [SPARK-29268](https://issues.apache.org/jira/browse/SPARK-29268).

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 3]
```diff
@@ -145,6 +145,20 @@ private[hive] class HiveClientImpl(
         warehouseDir.foreach { dir =>
           ret.getConf.setVar(ConfVars.METASTOREWAREHOUSE, dir)
         }
+        // Since in jdk11, when HiveClient's withHiveSate, it will set ThreadLocal SessionState,
+        // In HiveThriftServer2, it will call HiveUtils.newClientForExecution() to get a client
+        // for Execution, then that method will trigger here to execute, in that case we can't reset
+        // ret.getConf's ClassLoader.
+        if (HiveUtils.isCliSessionState) {
```
Comment (AngersZhuuuu) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: @srowen @wangyum  

Some special point for jdk11.  Need to add more explanation？

**Comment by AngersZhuuuu** [comment]
@dongjoon-hyun 
Passed jdk11 test,  I have changed the title, thanks for trigger build for jdk8 & hadoop3.2.

**Comment by wangyum** [comment]
retest this please

**Comment by wangyum** [comment]
retest this please

**Review by srowen — COMMENTED** [review]
Looks plausible. @wangyum ?

**Review by wangyum — APPROVED** [review]
Thank you for ping me @srowen. The change LGTM.

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**